### ISSUE TITLE
Build fails for wasm32-unknown-unknown target

### ISSUE THREAD

**Issue by sallar** [briansmith/ring#657]
Hi,

I'm trying to use Ring in WebAssembly, but when I try to build it for `wasm32-unknown-unknown` I get this error:

```
› cargo build --target wasm32-unknown-unknown
   Compiling unicode-xid v0.1.0
   Compiling serde v1.0.62
   Compiling wasm-bindgen-shared v0.2.11
   Compiling itoa v0.4.1
   Compiling dtoa v0.4.2
   Compiling cc v1.0.15
   Compiling libc v0.2.41
   Compiling untrusted v0.6.1
   Compiling proc-macro2 v0.4.3
   Compiling ring v0.13.0-alpha3
   Compiling quote v0.6.2
   Compiling syn v0.14.0
error: failed to run custom build command for `ring v0.13.0-alpha3`
process didn't exit successfully: `/Users/sallar/Projects/buttercup/wasmcane/target/debug/build/ring-66839d4f06759801/build-script-build` (exit code: 101)
--- stdout
Apple_PubSub_Socket_Render: /private/tmp/com.apple.launchd.MhcIwTZlPI/Render
CARGO: /Users/sallar/.rustup/toolchains/nightly-x86_64-apple-darwin/bin/cargo
CARGO_CFG_DEBUG_ASSERTIONS:
CARGO_CFG_PROC_MACRO:
CARGO_CFG_TARGET_ARCH: wasm32
CARGO_CFG_TARGET_ENDIAN: little
CARGO_CFG_TARGET_ENV:
CARGO_CFG_TARGET_HAS_ATOMIC: 16,32,8,ptr
CARGO_CFG_TARGET_OS: unknown
CARGO_CFG_TARGET_POINTER_WIDTH: 32
CARGO_CFG_TARGET_VENDOR: unknown
CARGO_FEATURE_DEFAULT: 1
CARGO_FEATURE_DEV_URANDOM_FALLBACK: 1
CARGO_FEATURE_USE_HEAP: 1
CARGO_HOME: /Users/sallar/.cargo
CARGO_MAKEFLAGS: --jobserver-fds=3,8 -j --jobserver-auth=3,8 -j
CARGO_MANIFEST_DIR: /Users/sallar/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.13.0-alpha3
CARGO_MANIFEST_LINKS: ring-asm
CARGO_PKG_AUTHORS: Brian Smith <brian@briansmith.org>
CARGO_PKG_DESCRIPTION: Safe, fast, small crypto using Rust.
CARGO_PKG_HOMEPAGE:
CARGO_PKG_NAME: ring
CARGO_PKG_VERSION: 0.13.0-alpha3
CARGO_PKG_VERSION_MAJOR: 0
CARGO_PKG_VERSION_MINOR: 13
CARGO_PKG_VERSION_PATCH: 0
CARGO_PKG_VERSION_PRE: alpha3
CLICOLOR: true
COLORFGBG: 15;0
COLORTERM: truecolor
COMMAND_MODE: unix2003
DEBUG: true
DYLD_LIBRARY_PATH: /Users/sallar/Projects/buttercup/wasmcane/target/debug/deps:/Users/sallar/.rustup/toolchains/nightly-x86_64-apple-darwin/lib:/Users/sallar/.rustup/toolchains/nightly-x86_64-apple-darwin/lib
HOME: /Users/sallar
HOST: x86_64-apple-darwin
ITERM_PROFILE: Sallar / zsh
ITERM_SESSION_ID: w0t0p0:267EC4C8-76AC-45D6-A280-2C37C9654288
LC_CTYPE: UTF-8
LOGNAME: sallar
LSCOLORS: exfxcxdxbxegedabagacad
NUM_JOBS: 8
OLDPWD: /Users/sallar/Projects/buttercup
OPT_LEVEL: 0
OUT_DIR: /Users/sallar/Projects/buttercup/wasmcane/target/wasm32-unknown-unknown/debug/build/ring-8e10f066167cde5e/out
PATH: /Users/sallar/.cargo/bin:/Users/sallar/.cargo/bin:/Users/sallar/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
PROFILE: debug
PROJECTS: /Users/sallar/Projects
PROMPT:
$(battery_status)in $(directory_name) $(git_dirty)$(need_push)
›
PWD: /Users/sallar/Projects/buttercup/wasmcane
RPROMPT: %{%}%{%}
RUSTC: rustc
RUSTDOC: rustdoc
RUSTUP_HOME: /Users/sallar/.rustup
RUSTUP_TOOLCHAIN: nightly-x86_64-apple-darwin
RUST_RECURSION_COUNT: 1
SECURITYSESSIONID: 186b1
SHELL: /usr/local/bin/zsh
SHLVL: 1
SSH_AUTH_SOCK: /private/tmp/com.apple.launchd.ZfUlNApE83/Listeners
TARGET: wasm32-unknown-unknown
TERM: xterm-256color
TERM_PROGRAM: iTerm.app
TERM_PROGRAM_VERSION: 3.1.6beta5
TERM_SESSION_ID: w0t0p0:267EC4C8-76AC-45D6-A280-2C37C9654288
TMPDIR: /var/folders/jc/c9wh_v4178l6tchxbklvjtgc0000gn/T/
USER: sallar
XPC_FLAGS: 0x0
XPC_SERVICE_NAME: 0
ZSH: /Users/sallar/.dotfiles
_: /Users/sallar/.cargo/bin/cargo
__CF_USER_TEXT_ENCODING: 0x1F5:0x0:0x0

--- stderr
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', libcore/option.rs:345:21
note: Run with `RUST_BACKTRACE=1` for a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed
```

I even updated to the latest alpha, but didn't work.

Thank you :)

**Comment by eucpp** [comment]
I got same for wasm32-unknown-emscripten

**Comment by briansmith** [comment]
> I'm trying to use Ring in WebAssembly

Why are you trying to use *ring* in WebAssembly?

*ring*'s performance and side-channel attack resistance come largely from its hand-coded native (x86_64, x86, ARM, Aarch64) assembly language code, which can't work in WebAssembly. Web browsers expose a standard crypto API to web content, and web content should generally use that crypto API, the Web Crypto API, instead of providing its own.

*ring* can't simply wrap the Web Crypto API because the Web Crypto API is asynchronous and *ring*'s API is synchronous.

**Comment by sallar** [comment]
@briansmith I have anyway fixed my issue by not using ring and instead using the new RustCrypto packages that are being developed since those are all WASM-ready.

But anyway the point of using Ring in WebAssembly was using the same stack and functions across all of our products in Browser, ElectronJS, ReactNative etc. For ReactNative we are not using WASM and instead compile the rust code to staticlib.

**Comment by briansmith** [comment]
Can you implement crypto in WebAssembly safely? AFAICT WebAssembly doesn't make any guarantees that are necessary for cryptography w.r.t. side channel resistance. That's why I recommend people use the web crypto API instead, within a browser.

**Comment by sallar** [comment]
@briansmith well all the https://github.com/RustCrypto packages are supposed to be WASM ready AFAIK

**Comment by briansmith** [comment]
> Can you implement crypto in WebAssembly safely? AFAICT WebAssembly doesn't make any guarantees that are necessary for cryptography w.r.t. side channel resistance. That's why I recommend people use the web crypto API instead, within a browser.

So, after researching, it seems the web crypto API isn't really a workable solution for many applications. So, I am open to adding WebAssembly support to *ring*. My expectation is that we'd do it by writing pure Rust implementations of the algorithms for which we have only assembly language implementations now. Anybody who is interested in potentially sponsoring that work should email me at brian@briansmith.org.

**Comment by josephlr** [comment]
It's worth noting that even a pure Rust implementation still has to get it's randomness from somewhere. The [`getrandom` crate](https://github.com/rust-random/getrandom) on `wasm32-unknown-unknown` uses either:
  - [`window.crypto.getRandomValues`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) when running in the browser ([Rust code](https://github.com/rust-random/getrandom/blob/0b87a0974b75357d55d6ec2c0358a81371294716/src/wasm32_stdweb.rs#L84))
  - [`crypto.randomBytes`](https://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback) when running under Node ([Rust code](https://github.com/rust-random/getrandom/blob/0b87a0974b75357d55d6ec2c0358a81371294716/src/wasm32_stdweb.rs#L94))

Both of these require calling out to JavaScript to call the browser APIs. So if we have to call out to the browser anyway, we might as well use the browser's built-in crypto.

[`wasm32-wasi`](https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-overview.md) is another matter. Here we have [`__wasi_random_get()`](https://github.com/CraneStation/wasmtime/blob/26cbbaab4140867c60d345a0f040cf4178bbbfdc/docs/WASI-api.md#random_get).

**Comment by vhnatyk** [comment]
glad ring going to be on wasm soon

**Comment by briansmith** [comment]
PR #863 implements SHA-2 in Rust to complement the Rust SHA-1 implementation. It should kick in automatically for the wasm32-unknown-unknown target if/when the build support for wasm32-unknown-unknown is added. PTAL if you are interested in a wasm32 port.

**Comment by briansmith** [comment]
Closing this since the build doesn't fail for wasm32-unknown-unknown. Not all the functionality is implemented for this target yet, but that's a separate issue.

**Comment by briansmith** [comment]
> Closing this since the build doesn't fail for wasm32-unknown-unknown. Not all the functionality is implemented for this target yet, but that's a separate issue.

Just to update this, AFAICT all the functionality of *ring* have been available for wasm32-unknown-unknown since 0.17.0.
